easybuild/easyblocks/n/nvhpc.py

```diff
@@ -0,0 +1,201 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with EasyBuild.  If not, see <http://www.gnu.org/licenses/>.
+##
+"""
+EasyBuild support for installing NVIDIA HPC SDK compilers, based on the easyblock for PGI compilers
+
+@author: Bart Oldeman (McGill University, Calcul Quebec, Compute Canada)
+@author: Damian Alvarez (Forschungszentrum Juelich)
+@author: Andreas Herten (Forschungszentrum Juelich)
+"""
+import os
+import fileinput
+import re
+import stat
+import sys
+
+import easybuild.tools.environment as env
+from distutils.version import LooseVersion
+from easybuild.easyblocks.generic.packedbinary import PackedBinary
+from easybuild.framework.easyconfig import CUSTOM
+from easybuild.tools.filetools import adjust_permissions, write_file
+from easybuild.tools.run import run_cmd
+from easybuild.tools.modules import get_software_root
+
+
+# contents for siterc file to make PGI/NVHPC pick up $LIBRARY_PATH
+# cfr. https://www.pgroup.com/support/link.htm#lib_path_ldflags
+SITERC_LIBRARY_PATH = """
+# get the value of the environment variable LIBRARY_PATH
+variable LIBRARY_PATH is environment(LIBRARY_PATH);
+
+# split this value at colons, separate by -L, prepend 1st one by -L
+variable library_path is
+default($if($LIBRARY_PATH,-L$replace($LIBRARY_PATH,":", -L)));
+
+# add the -L arguments to the link line
+append LDLIBARGS=$library_path;
+
+# also include the location where libm & co live on Debian-based systems
+# cfr. https://github.com/easybuilders/easybuild-easyblocks/pull/919
+append LDLIBARGS=-L/usr/lib/x86_64-linux-gnu;
+"""
+
+
+class EB_NVHPC(PackedBinary):
+    """
+    Support for installing the NVIDIA HPC SDK (NVHPC) compilers
+    """
+
+    @staticmethod
+    def extra_options():
+        extra_vars = {
+            'default_cuda_version': ["10.1", "CUDA Version to be used as default (10.2, 11.0, ...)", CUSTOM],
+            'compute_capability':   ["70", "Compute Capability (70, 80, ...)", CUSTOM],
+            'module_byo_compilers': [False, "BYO Compilers: Remove compilers from module", CUSTOM],
+            'module_nvhpc_own_mpi': [False, "Add NVHPC's packaged OpenMPI to module", CUSTOM],
+            'module_add_math_libs': [False, "Add NVHPC's math libraries to module", CUSTOM],
+            'module_add_profilers': [False, "Add NVHPC's NVIDIA Profilers to module", CUSTOM],
+            'module_add_nccl':      [False, "Add NVHPC's NCCL library to module", CUSTOM],
+            'module_add_nvshmem':   [False, "Add NVHPC's NVSHMEM library to module", CUSTOM],
+            'module_add_cuda':      [False, "Add NVHPC's CUDA to module", CUSTOM]
+        }
+        return PackedBinary.extra_options(extra_vars)
+
+    def __init__(self, *args, **kwargs):
+        """Easyblock constructor, define custom class variables specific to NVHPC."""
+        super(EB_NVHPC, self).__init__(*args, **kwargs)
+
+        self.nvhpc_install_subdir = os.path.join('Linux_x86_64', self.version)
```

Comment:
Does NVHPC also support non-x86_64 systems?

If so, we should use `get_cpu_architecture()` from `easybuild.tools.systemtools` here (and potentially do a mapping), like we do in the [CUDA easyblock](https://github.com/easybuilders/easybuild-easyblocks/blob/5e2a7577f2ddcc04b0364b577b9d84b520842540/easybuild/easyblocks/c/cuda.py#L64).

Reply:
According to their website, they support "Linux x86_64", "Linux OpenPOWER", "Linux Arm Server", "Windows x86 (eventually)".

I can have a quick check if the sub-directories are named in a sensible manner, but I don't have time (or capabilitie) to test all versions…

Reply:
I would suggest to leave that as "room for improvement" We do not have any production system to test that, so I think it makes sense to get this merged, and then when someone needs it in one of these systems, it could be improved by people with access to those systems.

Reply:
OK, makes sense to keep this hardcoded now and revisit it later.

Reply:
I've added a note to the code in 7937f130.